STDDEV doesn't work for floats or decimals

```sql
CREATE VIEW slot_stats AS SELECT
    STDDEV(x) / AVG(x) AS y
FROM ...
```

where `x` is `DOUBLE` causes. 

```
error[E0277]: the trait bound `dbsp::algebra::F64: num_traits::int::PrimInt` is not satisfied
   --> project018e83eb-b2c0-7c0d-8538-0b749c8fc644/src/main.rs:617:42
    |
617 | ...   Tup15::new(a.0, a.1, sqrt_dN(div_nullNN(minus_dN_dN(a.2.2, div_dN_dN(times_dN_dN(a.2.0, a.2.0), a.2.1)), minus_dN_dN(a.2.1, Some(F6...
    |                                    ^^^^^^^^^^ the trait `num_traits::int::PrimInt` is not implemented for `dbsp::algebra::F64`
    |
    = help: the following other types implement trait `num_traits::int::PrimInt`:
              isize
              i8
              i16
              i32
              i64
              i128
              usize
              u8
            and 4 others
note: required by a bound in `sqllib::operators::div_nullNN`

```

--- a/crates/dbsp/src/algebra/mod.rs
+++ b/crates/dbsp/src/algebra/mod.rs
@@ -20,7 +20,11 @@ pub use zset::{
     OrdZSetFactories, ZBatch, ZBatchReader, ZCursor, ZSet, ZSetReader, ZTrace, ZWeight,
 };
 
-use rust_decimal::Decimal;
+use rust_decimal::{
+    prelude::One,
+    prelude::Zero,
+    Decimal,
+};
 use size_of::SizeOf;
 use std::{
     marker::PhantomData,
@@ -73,6 +77,18 @@ impl_has_zero! {
     isize,
 }
 
+impl HasZero for Decimal {
+    #[inline]
+    fn is_zero(&self) -> bool {
+        Zero::is_zero(self)
+    }
+
+    #[inline]
+    fn zero() -> Self {
+        Zero::zero()
+    }
+}
+
 impl<T> HasZero for Option<T> {
     #[inline]
     fn is_zero(&self) -> bool {
@@ -120,6 +136,13 @@ macro_rules! impl_has_one {
             }
         )*
     };
+}
+
+impl HasOne for Decimal {
+    #[inline]
+    fn one() -> Self {
+        One::one()
+    }
 }
 
 impl_has_one! {
@@ -279,6 +302,24 @@ where
 
 /// `MulByRef<isize>`
 
+impl MulByRef<isize> for i8 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &isize) -> Self::Output {
+        (*self as isize * w) as Self
+    }
+}
+
+impl MulByRef<isize> for i16 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &isize) -> Self::Output {
+        (*self as isize * w) as Self
+    }
+}
+
 impl MulByRef<isize> for i32 {
     type Output = Self;
 
@@ -344,6 +385,24 @@ impl MulByRef<isize> for Decimal {
 
 /////////// `MulByRef<i64>`
 
+impl MulByRef<i64> for i8 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &i64) -> Self::Output {
+        (*self as i64 * w) as Self
+    }
+}
+
+impl MulByRef<i64> for i16 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &i64) -> Self::Output {
+        (*self as i64 * w) as Self
+    }
+}
+
 impl MulByRef<i64> for i32 {
     type Output = Self;
 
@@ -408,6 +467,24 @@ impl MulByRef<i64> for Decimal {
 }
 
 /////////// `MulByRef<i32>`
+
+impl MulByRef<i32> for i8 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &i32) -> Self::Output {
+        (*self as i32 * w) as Self
+    }
+}
+
+impl MulByRef<i32> for i16 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_by_ref(&self, w: &i32) -> Self::Output {
+        (*self as i32 * w) as Self
+    }
+}
 
 impl MulByRef<i32> for i64 {
     type Output = Self;
@@ -476,6 +553,8 @@ impl MulByRef<i32> for Decimal {
 
 trait OptionWeightType {}
 impl OptionWeightType for isize {}
+impl OptionWeightType for i8 {}
+impl OptionWeightType for i16 {}
 impl OptionWeightType for i32 {}
 impl OptionWeightType for i64 {}
 impl OptionWeightType for f32 {}

--- a/crates/dbsp/src/algebra/mod.rs
+++ b/crates/dbsp/src/algebra/mod.rs
@@ -646,9 +646,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_integer_tests() {
-        assert_eq!(0, i64::zero());
-        assert_eq!(1, i64::one());
-        let two = i64::one().add_by_ref(&i64::one());
+        assert_eq!(0, <i64 as algebra::HasZero>::zero());
+        assert_eq!(1, <i64 as algebra::HasOne>::one());
+        let two = <i64 as algebra::HasOne>::one().add_by_ref(&<i64 as algebra::HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));
@@ -656,9 +656,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_isize_tests() {
-        assert_eq!(0, isize::zero());
-        assert_eq!(1, isize::one());
-        let two = isize::one().add_by_ref(&isize::one());
+        assert_eq!(0, <isize as algebra::HasZeo>::zero());
+        assert_eq!(1, <isize as algebra::HasOne>::one());
+        let two = <isize as algebra::HasOne>::one().add_by_ref(&<isize as algebra::HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));
@@ -666,9 +666,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_i64_tests() {
-        assert_eq!(0, i64::zero());
-        assert_eq!(1, i64::one());
-        let two = i64::one().add_by_ref(&i64::one());
+        assert_eq!(0, <i64 as algebra::HasZero>::zero());
+        assert_eq!(1, <i64 as algebra::HasOne>::one());
+        let two = <i64 as algebra::HasOne>::one().add_by_ref(&<i64 as algebra::HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));

--- a/crates/dbsp/src/algebra/mod.rs
+++ b/crates/dbsp/src/algebra/mod.rs
@@ -646,9 +646,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_integer_tests() {
-        assert_eq!(0, <i64 as algebra::HasZero>::zero());
-        assert_eq!(1, <i64 as algebra::HasOne>::one());
-        let two = <i64 as algebra::HasOne>::one().add_by_ref(&<i64 as algebra::HasOne>::one());
+        assert_eq!(0, <i64 as HasZero>::zero());
+        assert_eq!(1, <i64 as HasOne>::one());
+        let two = <i64 as HasOne>::one().add_by_ref(&<i64 as HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));
@@ -656,9 +656,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_isize_tests() {
-        assert_eq!(0, <isize as algebra::HasZeo>::zero());
-        assert_eq!(1, <isize as algebra::HasOne>::one());
-        let two = <isize as algebra::HasOne>::one().add_by_ref(&<isize as algebra::HasOne>::one());
+        assert_eq!(0, <isize as HasZero>::zero());
+        assert_eq!(1, <isize as HasOne>::one());
+        let two = <isize as HasOne>::one().add_by_ref(&<isize as HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));
@@ -666,9 +666,9 @@ mod integer_ring_tests {
 
     #[test]
     fn fixed_i64_tests() {
-        assert_eq!(0, <i64 as algebra::HasZero>::zero());
-        assert_eq!(1, <i64 as algebra::HasOne>::one());
-        let two = <i64 as algebra::HasOne>::one().add_by_ref(&<i64 as algebra::HasOne>::one());
+        assert_eq!(0, <i64 as HasZero>::zero());
+        assert_eq!(1, <i64 as HasOne>::one());
+        let two = <i64 as HasOne>::one().add_by_ref(&<i64 as HasOne>::one());
         assert_eq!(2, two);
         assert_eq!(-2, two.neg_by_ref());
         assert_eq!(-4, two.mul_by_ref(&two.neg_by_ref()));

--- a/crates/dbsp/src/algebra/mod.rs
+++ b/crates/dbsp/src/algebra/mod.rs
@@ -20,11 +20,7 @@ pub use zset::{
     OrdZSetFactories, ZBatch, ZBatchReader, ZCursor, ZSet, ZSetReader, ZTrace, ZWeight,
 };
 
-use rust_decimal::{
-    prelude::One,
-    prelude::Zero,
-    Decimal,
-};
+use rust_decimal::{prelude::One, prelude::Zero, Decimal};
 use size_of::SizeOf;
 use std::{
     marker::PhantomData,

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -590,8 +590,6 @@ public class ToRustInnerVisitor extends InnerVisitor {
             this.builder.append(" - 1)");
             return VisitDecision.STOP;
         } else if (expression.operation == DBSPOpcode.DIV_NULL) {
-            if (expression.left.getType().is(DBSPTypeDecimal.class))
-                this.builder.append("decimal_");
             this.builder.append("div_null")
                     .append(expression.left.getType().nullableSuffix())
                     .append(expression.right.getType().nullableSuffix())

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java
@@ -417,7 +417,7 @@ public class CalciteToDBSPCompiler extends RelVisitor
         DBSPExpression[] keys = new DBSPExpression[fields.cardinality()];
         int next = 0;
         for (int index : fields) {
-            keys[next] = t.deref().field(index).applyCloneIfNeeded().cast(slice.getFieldType(index));
+            keys[next] = t.deepCopy().deref().field(index).applyCloneIfNeeded().cast(slice.getFieldType(index));
             next++;
         }
         return new DBSPTupleExpression(keys);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
@@ -49,6 +49,7 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPStringLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPTimeLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.IHasZero;
 import org.dbsp.sqlCompiler.ir.type.IsNumericType;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDate;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDecimal;
@@ -379,7 +380,7 @@ public class Simplify extends InnerRewriteVisitor {
         } else if (expression.operation.equals(DBSPOpcode.ADD)) {
             if (left.is(DBSPLiteral.class)) {
                 DBSPLiteral leftLit = left.to(DBSPLiteral.class);
-                IsNumericType iLeftType = leftType.to(IsNumericType.class);
+                IHasZero iLeftType = leftType.to(IHasZero.class);
                 if (iLeftType.isZero(leftLit)) {
                     result = right;
                 } else if (leftLit.isNull) {
@@ -387,7 +388,7 @@ public class Simplify extends InnerRewriteVisitor {
                 }
             } else if (right.is(DBSPLiteral.class)) {
                 DBSPLiteral rightLit = right.to(DBSPLiteral.class);
-                IsNumericType iRightType = rightType.to(IsNumericType.class);
+                IHasZero iRightType = rightType.to(IHasZero.class);
                 if (iRightType.isZero(rightLit)) {
                     result = left;
                 } else if (rightLit.isNull) {

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/OptimizeDistinctVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/OptimizeDistinctVisitor.java
@@ -39,7 +39,7 @@ public class OptimizeDistinctVisitor extends CircuitCloneVisitor {
         // distinct (distinct) = distinct
         DBSPOperator input = this.mapped(distinct.input());
         if (input.is(DBSPStreamDistinctOperator.class)) {
-            this.map(distinct, input);
+            this.map(distinct, input, false);
             return;
         }
         if (input.is(DBSPStreamJoinOperator.class) ||

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IHasZero.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IHasZero.java
@@ -1,0 +1,11 @@
+package org.dbsp.sqlCompiler.ir.type;
+
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
+
+/** A type that has a Zero value */
+public interface IHasZero {
+    DBSPLiteral getZero();
+    default boolean isZero(DBSPLiteral literal) {
+        return this.getZero().sameValue(literal);
+    }
+}

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsNumericType.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsNumericType.java
@@ -25,18 +25,11 @@ package org.dbsp.sqlCompiler.ir.type;
 
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 
-/**
- * Interface implemented by numeric types.
- */
-public interface IsNumericType {
-    default String getRustString() { return this.toString(); }
-    DBSPLiteral getZero();
+/** Interface implemented by numeric types. */
+public interface IsNumericType extends IHasZero {
     DBSPLiteral getOne();
     DBSPLiteral getMaxValue();
     DBSPLiteral getMinValue();
-    default boolean isZero(DBSPLiteral literal) {
-        return this.getZero().sameValue(literal);
-    }
     default boolean isOne(DBSPLiteral literal) {
         return this.getOne().sameValue(literal);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMillisInterval.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMillisInterval.java
@@ -29,19 +29,18 @@ import org.dbsp.sqlCompiler.compiler.visitors.inner.InnerVisitor;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPIntervalMillisLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.IHasZero;
 import org.dbsp.sqlCompiler.ir.type.IsDateType;
-import org.dbsp.sqlCompiler.ir.type.IsNumericType;
-import org.dbsp.sqlCompiler.compiler.errors.UnsupportedException;
 
 import java.util.Objects;
 
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.INTERVAL_SHORT;
 
-/**
- * Models the SQL Interval type for days-seconds.
- * Always stores the interval value in milliseconds.
- */
-public class DBSPTypeMillisInterval extends DBSPTypeBaseType implements IsNumericType, IsDateType {
+/** Models the SQL Interval type for days-seconds.
+ * Always stores the interval value in milliseconds. */
+public class DBSPTypeMillisInterval
+        extends DBSPTypeBaseType
+        implements IsDateType, IHasZero {
     public DBSPTypeMillisInterval(CalciteObject node, boolean mayBeNull) {
         super(node, INTERVAL_SHORT, mayBeNull);
     }
@@ -69,27 +68,12 @@ public class DBSPTypeMillisInterval extends DBSPTypeBaseType implements IsNumeri
 
     @Override
     public DBSPLiteral defaultValue() {
-        return this.getZero();
+        return new DBSPIntervalMillisLiteral(0, this.mayBeNull);
     }
 
     @Override
     public DBSPLiteral getZero() {
         return new DBSPIntervalMillisLiteral(0, this.mayBeNull);
-    }
-
-    @Override
-    public DBSPLiteral getOne() {
-        throw new UnsupportedException(this.getNode());
-    }
-
-    @Override
-    public DBSPLiteral getMaxValue() {
-        throw new UnsupportedException(this.getNode());
-    }
-
-    @Override
-    public DBSPLiteral getMinValue() {
-        throw new UnsupportedException(this.getNode());
     }
 
     @Override

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMonthsInterval.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMonthsInterval.java
@@ -30,17 +30,16 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPIntervalMonthsLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.DBSPTypeCode;
+import org.dbsp.sqlCompiler.ir.type.IHasZero;
 import org.dbsp.sqlCompiler.ir.type.IsDateType;
-import org.dbsp.sqlCompiler.ir.type.IsNumericType;
-import org.dbsp.sqlCompiler.compiler.errors.UnsupportedException;
 
 import java.util.Objects;
 
-/**
- * Models the SQL Interval type for months-years.
- * Always stores the interval value in days.
- */
-public class DBSPTypeMonthsInterval extends DBSPTypeBaseType implements IsNumericType, IsDateType {
+/** Models the SQL Interval type for months-years.
+ * Always stores the interval value in days. */
+public class DBSPTypeMonthsInterval
+        extends DBSPTypeBaseType
+        implements IsDateType, IHasZero {
     public DBSPTypeMonthsInterval(CalciteObject node, boolean mayBeNull) {
         super(node, DBSPTypeCode.INTERVAL_LONG, mayBeNull);
     }
@@ -68,27 +67,12 @@ public class DBSPTypeMonthsInterval extends DBSPTypeBaseType implements IsNumeri
 
     @Override
     public DBSPLiteral defaultValue() {
-        return this.getZero();
+        return new DBSPIntervalMonthsLiteral(0, this.mayBeNull);
     }
 
     @Override
     public DBSPLiteral getZero() {
         return new DBSPIntervalMonthsLiteral(0, this.mayBeNull);
-    }
-
-    @Override
-    public DBSPLiteral getOne() {
-        throw new UnsupportedException(this.getNode());
-    }
-
-    @Override
-    public DBSPLiteral getMaxValue() {
-        throw new UnsupportedException(this.getNode());
-    }
-
-    @Override
-    public DBSPLiteral getMinValue() {
-        throw new UnsupportedException(this.getNode());
     }
 
     @Override

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/OtherTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/OtherTests.java
@@ -497,6 +497,21 @@ public class OtherTests extends BaseSQLTests implements IWritesLogs {
     }
 
     @Test
+    public void testIOT() throws IOException, InterruptedException, SQLException {
+        // Iot code from different repository checked out in a specific place
+        final String iotSql = "../../../iot/iot.sql";
+        File sql = new File(iotSql);
+        if (!sql.exists())
+            return;
+        String script = Utilities.readFile(sql.toPath());
+        DBSPCompiler compiler = this.testCompiler();
+        compiler.options.languageOptions.throwOnError = true;
+        compiler.compileStatements(script);
+        CompilerCircuitStream ccs = new CompilerCircuitStream(compiler);
+        this.addRustTestCase("testIOT", ccs);
+    }
+
+    @Test
     public void testProjectFiles() throws IOException, InterruptedException, SQLException {
         // Compiles all the programs in the tests directory
         final String projectsDirectory = "../../demo/";

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/AggregateTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/AggregateTests.java
@@ -1,0 +1,167 @@
+package org.dbsp.sqlCompiler.compiler.sql.simple;
+
+import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
+import org.dbsp.sqlCompiler.compiler.sql.SqlIoTest;
+import org.junit.Test;
+
+public class AggregateTests extends SqlIoTest {
+    @Override
+    public void prepareInputs(DBSPCompiler compiler) {
+        compiler.compileStatements("""
+                CREATE TABLE T(
+                   B BIGINT,
+                   I INTEGER,
+                   S SMALLINT,
+                   T TINYINT,
+                   R REAL,
+                   D DOUBLE,
+                   E DECIMAL(3,2)
+                );
+                INSERT INTO T VALUES
+                   (NULL, NULL, NULL, NULL, NULL, NULL, NULL),
+                   (0, 0, 0, 0, 0, 0, 0),
+                   (1, 1, 1, 1, 1, 1, 1),
+                   (2, 2, 2, 2, 2, 2, 2);
+                """);
+    }
+
+    @Test
+    public void testAggregates() {
+        this.qs("""
+                SELECT COUNT(*), COUNT(B), COUNT(I), COUNT(S), COUNT(T), COUNT(R), COUNT(D), COUNT(E) FROM T;
+                 * | B | I | S | T | R | D | E
+                -------------------------------
+                 4 | 3 | 3 | 3 | 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT SUM(B), SUM(I), SUM(S), SUM(T), SUM(R), SUM(D), SUM(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 3 | 3 | 3 | 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT AVG(B), AVG(I), AVG(S), AVG(T), AVG(R), AVG(D), AVG(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT MIN(B), MIN(I), MIN(S), MIN(T), MIN(R), MIN(D), MIN(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 0 | 0 | 0 | 0 | 0 | 0 | 0
+                (3 row)
+                
+                SELECT MAX(B), MAX(I), MAX(S), MAX(T), MAX(R), MAX(D), MAX(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 2 | 2 | 2 | 2 | 2 | 2 | 2
+                (3 row)
+                
+                SELECT STDDEV(B), STDDEV(I), STDDEV(S), STDDEV(T), STDDEV(R), STDDEV(D), STDDEV(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT STDDEV_POP(B), STDDEV_POP(I), STDDEV_POP(S), STDDEV_POP(T), STDDEV_POP(R), STDDEV_POP(D), STDDEV_POP(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 0 | 0 | 0 | 0 | 0.8164966 | 0.816496580927726 | 0.82
+                (3 row)
+                
+                SELECT STDDEV_SAMP(B), STDDEV_SAMP(I), STDDEV_SAMP(S), STDDEV_SAMP(T), STDDEV_SAMP(R), STDDEV_SAMP(D), STDDEV_SAMP(E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT BIT_AND(B), BIT_AND(I), BIT_AND(S), BIT_AND(T) FROM T;
+                 B | I | S | T
+                ---------------
+                 0 | 0 | 0 | 0
+                (3 row)
+                
+                SELECT BIT_OR(B), BIT_OR(I), BIT_OR(S), BIT_OR(T) FROM T;
+                 B | I | S | T
+                ---------------
+                 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT BIT_XOR(B), BIT_XOR(I), BIT_XOR(S), BIT_XOR(T) FROM T;
+                 B | I | S | T
+                ---------------
+                 3 | 3 | 3 | 3
+                (3 row)""");
+    }
+
+    @Test
+    public void testDistinctAggregates() {
+        this.qs("""
+                SELECT COUNT(*), COUNT(DISTINCT B), COUNT(DISTINCT I), COUNT(DISTINCT S), COUNT(DISTINCT T), COUNT(DISTINCT R), COUNT(DISTINCT D), COUNT(DISTINCT E) FROM T;
+                 * | B | I | S | T | R | D | E
+                -------------------------------
+                 4 | 3 | 3 | 3 | 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT SUM(DISTINCT B), SUM(DISTINCT I), SUM(DISTINCT S), SUM(DISTINCT T), SUM(DISTINCT R), SUM(DISTINCT D), SUM(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 3 | 3 | 3 | 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT AVG(DISTINCT B), AVG(DISTINCT I), AVG(DISTINCT S), AVG(DISTINCT T), AVG(DISTINCT R), AVG(DISTINCT D), AVG(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT MIN(DISTINCT B), MIN(DISTINCT I), MIN(DISTINCT S), MIN(DISTINCT T), MIN(DISTINCT R), MIN(DISTINCT D), MIN(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 0 | 0 | 0 | 0 | 0 | 0 | 0
+                (3 row)
+                
+                SELECT MAX(DISTINCT B), MAX(DISTINCT I), MAX(DISTINCT S), MAX(DISTINCT T), MAX(DISTINCT R), MAX(DISTINCT D), MAX(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 2 | 2 | 2 | 2 | 2 | 2 | 2
+                (3 row)
+                
+                SELECT STDDEV(DISTINCT B), STDDEV(DISTINCT I), STDDEV(DISTINCT S), STDDEV(DISTINCT T), STDDEV(DISTINCT R), STDDEV(DISTINCT D), STDDEV(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT STDDEV_POP(DISTINCT B), STDDEV_POP(DISTINCT I), STDDEV_POP(DISTINCT S), STDDEV_POP(DISTINCT T), STDDEV_POP(DISTINCT R), STDDEV_POP(DISTINCT D), STDDEV_POP(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 0 | 0 | 0 | 0 | 0.8164966 | 0.816496580927726 | 0.82
+                (3 row)
+                
+                SELECT STDDEV_SAMP(DISTINCT B), STDDEV_SAMP(DISTINCT I), STDDEV_SAMP(DISTINCT S), STDDEV_SAMP(DISTINCT T), STDDEV_SAMP(DISTINCT R), STDDEV_SAMP(DISTINCT D), STDDEV_SAMP(DISTINCT E) FROM T;
+                 B | I | S | T | R | D | E
+                ---------------------------
+                 1 | 1 | 1 | 1 | 1 | 1 | 1
+                (3 row)
+                
+                SELECT BIT_AND(DISTINCT B), BIT_AND(DISTINCT I), BIT_AND(DISTINCT S), BIT_AND(DISTINCT T) FROM T;
+                 B | I | S | T
+                ---------------
+                 0 | 0 | 0 | 0
+                (3 row)
+                
+                SELECT BIT_OR(DISTINCT B), BIT_OR(DISTINCT I), BIT_OR(DISTINCT S), BIT_OR(DISTINCT T) FROM T;
+                 B | I | S | T
+                ---------------
+                 3 | 3 | 3 | 3
+                (3 row)
+                
+                SELECT BIT_XOR(DISTINCT B), BIT_XOR(DISTINCT I), BIT_XOR(DISTINCT S), BIT_XOR(DISTINCT T) FROM T;
+                 B | I | S | T
+                ---------------
+                 3 | 3 | 3 | 3
+                (3 row)""");
+    }
+}

--- a/sql-to-dbsp-compiler/lib/sqllib/src/casts.rs
+++ b/sql-to-dbsp-compiler/lib/sqllib/src/casts.rs
@@ -203,21 +203,16 @@ cast_function!(Time, Time, Time, Time);
 #[inline]
 pub fn cast_to_decimal_b(value: bool, precision: u32, scale: u32) -> Decimal {
     let result = if value {
-        Decimal::one()
+        <rust_decimal::Decimal as One>::one()
     } else {
-        Decimal::zero()
+        <rust_decimal::Decimal as Zero>::zero()
     };
     cast_to_decimal_decimal(result, precision, scale)
 }
 
 #[inline]
 pub fn cast_to_decimal_bN(value: Option<bool>, precision: u32, scale: u32) -> Decimal {
-    let result = if value.unwrap() {
-        Decimal::one()
-    } else {
-        Decimal::zero()
-    };
-    cast_to_decimal_decimal(result, precision, scale)
+    cast_to_decimal_b(value.unwrap(), precision, scale)
 }
 
 #[inline]
@@ -286,7 +281,7 @@ pub fn cast_to_decimal_s(value: String, precision: u32, scale: u32) -> Decimal {
 #[inline]
 pub fn cast_to_decimal_sN(value: Option<String>, precision: u32, scale: u32) -> Decimal {
     let result = match value {
-        None => Decimal::zero(),
+        None => <rust_decimal::Decimal as Zero>::zero(),
         Some(x) => x.trim().parse().unwrap(),
     };
     cast_to_decimal_decimal(result, precision, scale)
@@ -344,17 +339,17 @@ pub fn cast_to_decimalN_nullN(_value: Option<()>, _precision: u32, _scale: i32) 
 #[inline]
 pub fn cast_to_decimalN_b(value: bool, precision: u32, scale: u32) -> Option<Decimal> {
     let result = if value {
-        Some(Decimal::one())
+        Some(<rust_decimal::Decimal as One>::one())
     } else {
-        Some(Decimal::zero())
+        Some(<rust_decimal::Decimal as Zero>::zero())
     };
     set_ps(result, precision, scale)
 }
 
 #[inline]
 pub fn cast_to_decimalN_bN(value: Option<bool>, precision: u32, scale: u32) -> Option<Decimal> {
-    let result = value.map(|x| if x { Decimal::one() } else { Decimal::zero() });
-    set_ps(result, precision, scale)
+    let value = value?;
+    cast_to_decimalN_b(value, precision, scale)
 }
 
 #[inline]
@@ -404,23 +399,14 @@ pub fn cast_to_decimalN_fN(value: Option<F32>, precision: u32, scale: u32) -> Op
 
 #[inline]
 pub fn cast_to_decimalN_s(value: String, precision: u32, scale: u32) -> Option<Decimal> {
-    let result = match value.trim().parse() {
-        Err(_) => Some(Decimal::zero()),
-        Ok(x) => Some(x),
-    };
+    let result = Some(value.trim().parse().unwrap());
     set_ps(result, precision, scale)
 }
 
 #[inline]
 pub fn cast_to_decimalN_sN(value: Option<String>, precision: u32, scale: u32) -> Option<Decimal> {
-    let result = match value {
-        None => None,
-        Some(x) => match x.trim().parse() {
-            Err(_) => Some(Decimal::zero()),
-            Ok(y) => Some(y),
-        },
-    };
-    set_ps(result, precision, scale)
+    let value = value?;
+    cast_to_decimalN_s(value, precision, scale)
 }
 
 /////////// cast to double

--- a/sql-to-dbsp-compiler/lib/sqllib/src/operators.rs
+++ b/sql-to-dbsp-compiler/lib/sqllib/src/operators.rs
@@ -1,6 +1,6 @@
 use std::ops::{Add, Div, Mul, Sub};
 
-use dbsp::algebra::{F32, F64};
+use dbsp::algebra::{HasZero, F32, F64};
 use num::PrimInt;
 use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, ToPrimitive};
 
@@ -265,91 +265,39 @@ pub fn plus_u_u(left: usize, right: usize) -> usize {
 #[inline(always)]
 pub fn div_null__<T>(left: T, right: T) -> Option<T>
 where
-    T: PrimInt,
+    T: Div<Output = T> + HasZero,
 {
-    if Some(0) == right.to_isize() {
+    if right.is_zero() {
         None
     } else {
-        Some(left / right)
+        Some(left.div(right))
     }
 }
 
 #[inline(always)]
 pub fn div_nullN_<T>(left: Option<T>, right: T) -> Option<T>
 where
-    T: PrimInt,
+    T: Div<Output = T> + HasZero,
 {
     let left = left?;
-    if Some(0) == right.to_isize() {
-        None
-    } else {
-        Some(left / right)
-    }
+    div_null__::<T>(left, right)
 }
 
 #[inline(always)]
 pub fn div_null_N<T>(left: T, right: Option<T>) -> Option<T>
 where
-    T: PrimInt,
+    T: Div<Output = T> + HasZero,
 {
     let right = right?;
-    if Some(0) == right.to_isize() {
-        None
-    } else {
-        Some(left / right)
-    }
+    div_null__(left, right)
 }
 
 #[inline(always)]
 pub fn div_nullNN<T>(left: Option<T>, right: Option<T>) -> Option<T>
 where
-    T: PrimInt,
+    T: Div<Output = T> + HasZero,
 {
     let left = left?;
     let right = right?;
-    if Some(0) == right.to_isize() {
-        None
-    } else {
-        Some(left / right)
-    }
-}
-
-#[inline(always)]
-pub fn decimal_div_null(left: Decimal, right: Decimal) -> Option<Decimal> {
-    if Decimal::ZERO == right {
-        None
-    } else {
-        Some(left / right)
-    }
-}
-
-#[inline(always)]
-pub fn decimal_div_nullN_(left: Option<Decimal>, right: Decimal) -> Option<Decimal> {
-    let left = left?;
-    if Decimal::ZERO == right {
-        None
-    } else {
-        Some(left / right)
-    }
-}
-
-#[inline(always)]
-pub fn decimal_div_null_N(left: Decimal, right: Option<Decimal>) -> Option<Decimal> {
-    let right = right?;
-    if Decimal::ZERO == right {
-        None
-    } else {
-        Some(left / right)
-    }
-}
-
-#[inline(always)]
-pub fn decimal_div_nullNN(left: Option<Decimal>, right: Option<Decimal>) -> Option<Decimal> {
-    let left = left?;
-    let right = right?;
-    if Decimal::ZERO == right {
-        None
-    } else {
-        Some(left / right)
-    }
+    div_null__(left, right)
 }

